esphome/components/ags10/ags10.h

```diff
@@ -0,0 +1,160 @@
+
+class AGS10Component : public PollingComponent, public i2c::I2CDevice {
+ public:
+  /**
+   * Sets TVOC sensor.
+   */
+  void set_tvoc(sensor::Sensor *tvoc) { tvoc_ = tvoc; }
+
+  /**
+   * Sets version info sensor.
+   */
+  void set_version(sensor::Sensor *version) { version_ = version; }
+
+  /**
+   * Sets resistance info sensor.
+   */
+  void set_resistance(sensor::Sensor *resistance) { resistance_ = resistance; }
+
+  void setup() override;
+
+  void update() override;
+
+  void dump_config() override;
+
+  float get_setup_priority() const override { return setup_priority::DATA; }
+
+  /**
+   * Modifies target address of AGS10.
+   *
+   * New address is saved and takes effect immediately even after power-off.
+   */
+  bool new_i2c_address(uint8_t newaddress);
+
+  /**
+   * Sets zero-point with factory defaults.
+   */
+  bool set_zero_point_with_factory_defaults();
+
+  /**
+   * Sets zero-point with current sensor resistance.
+   */
+  bool set_zero_point_with_current_resistance();
+
+  /**
+   * Sets zero-point with the value.
+   */
+  bool set_zero_point_with(uint16_t value);
+
+ protected:
+  /**
+   *  TVOC.
+   */
+  sensor::Sensor *tvoc_{nullptr};
+
+  /**
+   * Firmvare version.
+   */
+  sensor::Sensor *version_{nullptr};
+
+  /**
+   * Resistance.
+   */
+  sensor::Sensor *resistance_{nullptr};
+
+  /**
+   *  Last operation error code.
+   */
+  enum ErrorCode {
+    NONE = 0,
+    COMMUNICATION_FAILED,
+    CRC_CHECK_FAILED,
+    ILLEGAL_STATUS,
+    UNSUPPORTED_UNITS,
+  } error_code_{NONE};
+
+  /**
+   *  Reads and returns value of TVOC.
+   */
+  optional<uint32_t> read_tvoc_();
+
+  /**
+   *  Reads and returns a firmware version of AGS10.
+   */
+  optional<uint8_t> read_version_();
+
+  /**
+   * Reads and returns the resistance of AGS10.
+   */
+  optional<uint32_t> read_resistance_();
+
+  /**
+   * Read, checks and returns data from the sensor.
+   */
+  template<size_t N> optional<std::array<uint8_t, N>> read_and_check_(uint8_t a_register);
+
+  /**
+   * Calculates CRC8 value.
+   *
+   * CRC8 calculation, initial value: 0xFF, polynomial: 0x31 (x8+ x5+ x4+1)
+   *
+   * @param[in] dat the data buffer
+   * @param num number of bytes in the buffer
+   */
+  template<size_t N> uint8_t calc_crc8_(std::array<uint8_t, N> dat, uint8_t num);
+};
+
+template<typename... Ts> class AGS10NewI2cAddressAction : public Action<Ts...> {
+ public:
+  AGS10NewI2cAddressAction(AGS10Component *parent) : parent_(parent) {}
+  TEMPLATABLE_VALUE(int, new_address)
+
+  void play(Ts... x) override { this->parent_->new_i2c_address(this->new_address_.value(x...)); }
+
+ protected:
+  AGS10Component *parent_;
+};
+
+enum AGS10SetZeroPointActionMode {
+  // Zero-point reset.
+  FACTORY_DEFAULT,
+  // Zero-point calibration with current resistance.
+  CURRENT_VALUE,
+  // Zero-point calibration with custom resistance.
+  CUSTOM_VALUE,
+};
+
+template<typename... Ts> class AGS10SetZeroPointAction : public Action<Ts...> {
```

Comment:
After adding this, you can remove the constructor and protected field
```suggestion
template<typename... Ts> class AGS10SetZeroPointAction : public Action<Ts...>, public Parented<AGS10Component> {
```

Reply:
Done

Reply:
I've to leave constructors, because the compilation error occurs without it:
```
src/main.cpp: In function 'void setup()':
src/main.cpp:1274:81: error: no matching function for call to 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(esphome::ags10::AGS10Component*&)'
   ags10_ags10newi2caddressaction = new ags10::AGS10NewI2cAddressAction<>(ags10_1);
                                                                                 ^
In file included from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction()'
 template<typename... Ts> class AGS10NewI2cAddressAction : public Action<Ts...>, public Parented<AGS10Component> {
                                ^~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ags10/ags10.h:116:32: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(const esphome::ags10::AGS10NewI2cAddressAction<>&)'
src/esphome/components/ags10/ags10.h:116:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'const esphome::ags10::AGS10NewI2cAddressAction<>&'
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(esphome::ags10::AGS10NewI2cAddressAction<>&&)'
src/esphome/components/ags10/ags10.h:116:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'esphome::ags10::AGS10NewI2cAddressAction<>&&'
src/main.cpp:1301:83: error: no matching function for call to 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(esphome::ags10::AGS10Component*&)'
   ags10_ags10newi2caddressaction_2 = new ags10::AGS10NewI2cAddressAction<>(ags10_1);
                                                                                   ^
In file included from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction()'
 template<typename... Ts> class AGS10NewI2cAddressAction : public Action<Ts...>, public Parented<AGS10Component> {
                                ^~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ags10/ags10.h:116:32: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(const esphome::ags10::AGS10NewI2cAddressAction<>&)'
src/esphome/components/ags10/ags10.h:116:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'const esphome::ags10::AGS10NewI2cAddressAction<>&'
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(esphome::ags10::AGS10NewI2cAddressAction<>&&)'
src/esphome/components/ags10/ags10.h:116:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'esphome::ags10::AGS10NewI2cAddressAction<>&&'
src/main.cpp:1329:79: error: no matching function for call to 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(esphome::ags10::AGS10Component*&)'
   ags10_ags10setzeropointaction = new ags10::AGS10SetZeroPointAction<>(ags10_1);
                                                                               ^
In file included from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction()'
 template<typename... Ts> class AGS10SetZeroPointAction : public Action<Ts...>, public Parented<AGS10Component> {
                                ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ags10/ags10.h:132:32: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(const esphome::ags10::AGS10SetZeroPointAction<>&)'
src/esphome/components/ags10/ags10.h:132:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'const esphome::ags10::AGS10SetZeroPointAction<>&'
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(esphome::ags10::AGS10SetZeroPointAction<>&&)'
src/esphome/components/ags10/ags10.h:132:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'esphome::ags10::AGS10SetZeroPointAction<>&&'
src/main.cpp:1357:81: error: no matching function for call to 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(esphome::ags10::AGS10Component*&)'
   ags10_ags10setzeropointaction_2 = new ags10::AGS10SetZeroPointAction<>(ags10_1);
                                                                                 ^
In file included from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction()'
 template<typename... Ts> class AGS10SetZeroPointAction : public Action<Ts...>, public Parented<AGS10Component> {
                                ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ags10/ags10.h:132:32: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(const esphome::ags10::AGS10SetZeroPointAction<>&)'
src/esphome/components/ags10/ags10.h:132:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'const esphome::ags10::AGS10SetZeroPointAction<>&'
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(esphome::ags10::AGS10SetZeroPointAction<>&&)'
src/esphome/components/ags10/ags10.h:132:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'esphome::ags10::AGS10SetZeroPointAction<>&&'
```

Reply:
Finally I managed to remove constructors.